crates/turborepo-lib/src/package_manager.rs

```diff
@@ -134,13 +152,202 @@ impl PackageManager {
+        let manager = match manager {
+            "npm" => Some(PackageManager::Npm),
+            "yarn" => Some(Self::detect_berry_or_yarn(&version)?),
+            "pnpm" => Some(Self::detect_pnpm6_or_pnpm(&version)?),
+            _ => None,
+        };
+
+        Ok(manager)
+    }
+
+    #[cfg(test)]
+    fn get_yarn_version(_: &AbsoluteSystemPathBuf) -> Result<Version> {
+        Ok(env::var("TEST_YARN_VERSION")?.parse()?)
+    }
+
+    #[cfg(not(test))]
+    fn get_yarn_version(project_directory: &AbsoluteSystemPathBuf) -> Result<Version> {
+        let output = Command::new("yarn")
+            .arg("--version")
+            .current_dir(&project_directory)
+            .output()?;
+        let yarn_version_output = String::from_utf8(output.stdout)?;
+        Ok(yarn_version_output.trim().parse()?)
+    }
+
+    fn detect_package_manager(base: &CommandBase) -> Result<PackageManager> {
+        let mut detected_package_managers = vec![];
+        let project_directory = AbsoluteSystemPathBuf::new(&base.repo_root)?;
+        let npm_lockfile = project_directory
+            .join_relative(RelativeSystemPathBuf::new("package-lock.json").unwrap());
+        if npm_lockfile.exists() {
+            detected_package_managers.push(PackageManager::Npm);
+        }
+
+        let pnpm_lockfile =
+            project_directory.join_relative(RelativeSystemPathBuf::new("pnpm-lock.yaml").unwrap());
+        if pnpm_lockfile.exists() {
+            detected_package_managers.push(PackageManager::Pnpm);
+        }
+
+        let yarn_lockfile =
+            project_directory.join_relative(RelativeSystemPathBuf::new("yarn.lock").unwrap());
+        if yarn_lockfile.exists() {
+            let version = Self::get_yarn_version(&project_directory)?;
+            detected_package_managers.push(Self::detect_berry_or_yarn(&version)?);
+        }
+
+        match detected_package_managers.len() {
+            0 => {
+                let url = base.ui.apply(
+                    UNDERLINE.apply_to("https://nodejs.org/api/packages.html#packagemanager"),
+                );
+                Err(anyhow!(
+                    "We did not find a package manager specified in your root package.json. \
+                     Please set the \"packageManager\" property in your root package.json ({url}) \
+                     or run `npx @turbo/codemod add-package-manager` in the root of your monorepo."
+                ))
+            }
+            1 => Ok(detected_package_managers.pop().unwrap()),
+            _ => Err(anyhow!(
+                "We detected multiple package managers in your repository: {}. Please remove one \
+                 of them.",
+                detected_package_managers.into_iter().join(", ")
+            )),
+        }
+    }
+
+    fn parse_package_manager_string(manager: &str) -> Result<(&str, &str)> {
+        let package_manager_pattern =
+            Regex::new(r"(?P<manager>npm|pnpm|yarn)@(?P<version>\d+\.\d+\.\d+(-.+)?)")?;
```

Comment:
How often do we detect package managers? I assumed it was once, hence the lack of OnceCell

Reply:
It's probably only called once per run now, but if package graph construction moves to daemon it might happen multiple times. Not a big deal if the `OnceCell` overhead is too expensive.